rfcs/2022-03-15-11851-ingest-opentelemtry-traces.md

```diff
@@ -0,0 +1,368 @@
+supported on both sides working on better interoperability on that particular common ground would likely be a better
+option.
+
+## Cross cutting concerns
+
+N/A
+
+## Scope
+
+### In scope
+
+- `opentelemetry_traces` source, with both http and grpc support
+- Support `opentelemetry_traces` source to `datadog_traces` sink forwarding by dealing with:
+  - Traces normalization to a single format inside Vector
+  - Conversion to/from this format in all traces sources/sinks
+- APM stats computation logic, with an implementation for the `opentelemetry_traces` sources, applicable for all traces
+  sources
+
+### Out of scope
+
+N/A
+
+## Pain
+
+- Avoid complex setup when ingesting traces, ultimately pointing every tracing lib directly to Vector should just work
+  out-of-the-box with minimal config.
+
+## Proposal
+
+### User Experience
+
+- User would point OpenTelemtry tracing lib directly to a local Vector deployement
+- Vector would be configured with a minimal config looking like:
+
+```yaml
+sources:
+  otlp_traces:
+    type: opentelemetry_traces
+    address: "[::]:8081"
+    mode: grpc
+
+sinks:
+  dd_trace:
+    type: datadog_traces
+    default_api_key: 12345678abcdef
+    inputs:
+     - otlp_traces
+```
+
+And it should just work.
+
+### Implementation
+
+- `opentelemetry_traces` or simply `opentelemetry` sources:
+  - The gRPC variant would use Tonic to spawn a gRPC server (like the `vector` source in its v2 variation) and directly
+    use the [offical gRPC service definitions][otlp-grpc-def].
+  - HTTP variant would use a Warp server and attempt to decode protobuf payloads, as per the [specification][otlp-http],
+    payloads are encoded using protobuf either in binary format or in JSON format ([Protobuf schemas][otlp-proto-def]).
+    All the expected behaviours regarding the kind of requests/responses code and sequence are clearly defined as well
+    as the default URL path (`/v1/traces` for traces).
+- Internal traces representation/normalization, two options are opened see [outstanding
+  questions](#outstanding-questions), but the consensus is leaning towards a new dedicated container that would:
+  - Move away from from the current implementation that relies on `LogEvent` to a dedicated container, the
+    implementation will stay  in [./lib/vector-core/src/event/trace.rs][current-trace-in-vector]
+  - Borrow most of its semantic from the Opentelemetry Traces format, the `TraceEvent` would then evolve toward a
+    concrete dedicated container and contains a slice of spans, the span being directly inspired by the Opentelemtry
+    specification.
+  - Some top-level information (Like trace ID, trace-wide tags/metrics, the original format)
+  - Trace would not get native `VrlTarget` representation anymore, there is a bigger discussion there that should
+    probably be adressed separately. As an interim measure few fields may be exposed (like trace ID & trace-wide
+    tags).
```

Comment:
My point was that I'm not exactly sure what should be exposed right now, the way I see it would be to at least allow vrl conditions to be evaluated (lets say on "well known" fields like the trace_id or tags/metadata) for filtering/routing and defer the big picture (as in traces vs. remap transform) to a subsequent RFC.

Reply:
Yeah, I can see that.